### ISSUE TITLE
Fix bundle.js for web arena

### ISSUE THREAD

**Issue by dmessing** [appirio-tech/topcoder-app#924]
Take a peek at this [medium post](https://medium.com/@taru.blm/investigating-a-slow-running-script-on-topcoder-82a422af3dfc#.eb2re7mcd)

------Copied from above MEdium Post)----
I wanted to try Topcoder web arena. Unfortunately, the very first practice problem I tried never loaded. I logged in/out and tried several browsers, but no luck.
I decided to debug. I can barely understand JavaScript, but I hoped it would at least give me an idea of what’s happening.
Chrome debugger paused inside this code inside bundle.js:
updateCoderPlacement = function (coders) {
        coders.forEach(function (item) {
            item.roomPlace = 1;
            coders.forEach(function (other) {
                if (item.userName !== other.userName &&
                        item.totalPoints < other.totalPoints) {
                    item.roomPlace += 1;
                }
                });
            });
        };
It looked quadratic in the size of coders, and coders had 67374 items. That would explain the problem, since it was trying to go through 4 billion iterations before loading the data.
But what could it be doing?
After re-reading the code a few times, I had to accept that it’s just trying to calculate the rank of each item in coders based on the value of .totalPoints.
I was really surprised that topcoder of all websites would have code like this. Not to mention, wouldn’t they notice that the website doesn’t load?
In any case, it’s obvious that we need a rank_array function, but how to do it in JavaScript? Surprisingly SO search yielded O(N²) solution as the top result (an answer with a more efficient solution is right after). Is it really common to write code like this on the front-end?
Anyway, here’s my attempt at translating into JavaScript the solution I would have done in python:
updateCoderPlacement = function (coders) {
    sorted_idx = coders
      .map(function(a, i) {
        return i; 
        
      })
      .sort(function(a, b) {
        return coders[b].totalPoints - coders[a].totalPoints; 
        
      });
      
    sorted_idx
      .map(function(idx, i) {
        coders[idx].roomPlace = i+1
        
      });
    };
I guess topcoder worked out fine in the end. Even though I didn’t get to practice their algorithm problems in their web arena (which never loaded), the website itself provided a fun way to practice JavaScript.
Edit: I just realized that the code is actually open source. If someone wants to, please go ahead and submit a PR to fix this. I’m not familiar with JavaScript or front end. But it does seem that the bug is in resolvers.js which then gets compiled (?) into bundle.js.

**Comment by dmessing** [comment]
@talesforce think we could get this one added to the hunt?  Love to get it 🚀

**Comment by dmessing** [comment]
@talesforce any updates? 🔍

**Comment by birdofpreyru** [comment]
@ajefts It looks to me the Arena code, which should be fixed, is in another repo?

**Comment by ajefts** [comment]
@birdofpreyru Yes, it's a different repo.  I believe you'll find it here:  https://github.com/appirio-tech/arena-web

The arena stuff will be a little trickier to work on since we haven't done development on it for a while and we'll likely need to do some work on dev environments and/or mocks.  But, let's give it a try and figure it out...

**Comment by birdofpreyru** [comment]
Actually, this issue is solved in the `dev` branch of that repo. And it seems there are multiple issues opened, which refer to this problem. People also say there, that the issue has been solved, but the updated was never deployed into production.

**Comment by birdofpreyru** [comment]
@ajefts Have you seen my previous comment? Just to be sure for future, do you see them only if I tag your handle explicitly?

**Comment by ajefts** [comment]
yes, please mention me on things I need to reply to.  I'll "watch" the other stuff, but won't have the bandwidth to closely follow every thread.

So, for this one...if it's already fixed in the dev repo...then this becomes a deployment effort.  We haven't deployed web arena in probably over a year.  Could we run something to validate that the dev repo is in a good state?  Maybe run a challenge that requires people to get the app running locally and then create tests to validate the latest changes?

@skyhit worked a lot on the arena web app and can probably provide some advice....

**Comment by birdofpreyru** [comment]
@ajefts Running such challenge in principle sounds like a good idea to me. The only thing, that we need to outreach to the people who do use the Arena, to participate and review it. Something like, they will have to deploy it locally and carefully test that all the features are working, capturing it all on video, and providing a list of tested/detected issues, and then reviewers, also familiar with the Arena, will check out submissions, declaring those who had achieved a better coverage of the everything to be the winners (using Subjective Review Card).

A technical concern I have is that, probably, those using Arena and willing to be reviewers in such challenge, they are not in review board for code challenges, so we'll have to include them.

What budget are you ready to dedicate to this?

**Comment by birdofpreyru** [comment]
@ajefts Also, looking at the list of issues opened in the Arena repo (208!!) it looks like there is a lots more staff which can be fixed. Should it also be considered to be a part of this Bug Bash, or should I focus on the main site only, at the moment? Also, can I count on copilot payments for co-piloting all this (I mean, I have volunteered to work on this primarily to get the experience of co-piloting, rather than hoping to get $$ for myself, but it is a nice to know in order to properly manage my time between this project and competing in other challenges)?

**Comment by ajefts** [comment]
For the arena testing, ideally we'll get some arena competitors to help test...but let's also get some automated tests written.  Part of the first challenge will simply be validating the environment.  Once we have it running well in dev, we can reach out to algo competitors to help test with real use cases.

As far as budget, don't worry for now. I'll tell you when/if we need to slow down.

For the 208+ issues in arena, let's focus on that first practice room performance issue first.  Then we'll need to triage the rest.  Focus on the other community apps first until we get to do that triage and prioritization.

Yes, you'll get copilot payments for this.  Ideally, you'll make the most $ by running challenges and engaging the community, but we'll also do task payments to you for additional work that isn't pertinent to challenges.  I like to do task payments and also milestone payments for meeting certain deliverables.

**Comment by Sande3p** [comment]
*I need topcoder arena VM* to fix the. I managed to setup the arena app. but authentication fails.
ref: https://gyazo.com/94bf7f0b0ac841b937fb15423fa152ac

**Comment by Sande3p** [comment]
Most probably VM is issued from topcoder direct (The same page from where tc contest were launched). But don't konw how it's issued.
This is a 2yr old mail from Topcoder VM https://gyazo.com/b04144c6be5b7966bf087bcc0667926d  . Not sure if this could help.

**Comment by rainerlopez** [comment]
Tkx so much @Sande3p, we are exactly trying to fix the local deployment at: https://github.com/appirio-tech/topcoder-app/issues/957
Take a look @birdofpreyru

**Comment by birdofpreyru** [comment]
Thanks a lot guys! Just wanna say, that more knowledgeable people, than I, at TopCoder are aware that we need VMs, or some other help with the backend part. Though, they need some time to verify that VMs they have are up-to-date with the latest changes in the code. So, probably, it is better to take it easier for the moment, and wait for further news from them.

**Comment by ajefts** [comment]
@skyhit Would you be able to help us with the dev environment on this so people can make changes to the web arena?

**Comment by ajefts** [comment]
The updated updateCoderPlacement function has been deployed.  This seems to have helped, but it looks like there are other issues that prevent the problems from loading sporadically.  So, making progress but still some more work to do here...
